New model interface

luv-bansal opened the pull request:
<!---
The PR description should include WHAT, WHY, HOW it does things and how this PR is tested.
-->

<!-- ### What -->
<!--
You should try to state the WHAT in PR title only.
Your PR title should describe in short WHAT this PR does, e.g. "[AA-1] Add new Clarifai cool feature".
Uncomment the `### What` section above, if you decide to add more details about WHAT this PR does.
For example, you may choose to describe the WHAT in more detail in the below cases.
* If the PR title is too short to include what your PR does;
* If your PR does more than one thing, the title may not have enough space to include everything. In this case, consider creating multiple PRs to separate work and make it easier for everybody to understand. Alternatively, mention a list of things what this PR does in this section.
-->



### Why
<!-- Why is this PR trying to accomplish said thing? This is useful to understand your intention why this code should be merged to master. -->
*

### How
<!-- How is this PR trying to accomplish said thing? This is useful to understand the details of the code. The code reviewer will check this section to assist them with code review. -->
*

### Tests
<!-- How is this PR tested? Write some tests to convince yourself and the reviewer that your code works. Sometimes, no tests are necessary, but we should always ask the question "Can I add some tests for this PR?". -->
*

### Notes
<!-- Add any additional notes here. -->
*

phatvo9 commented on clarifai/runners/utils/data_handler.py:
```diff
@@ -1,231 +1,121 @@
+    self.video = video
 
-class BaseDataHandler:
+  def to_proto(self, output):
```
why it has to have 'output' here? I think it should turn whatever it has to output proto?

deigen commented:
Nice --- really on the right track.  Things from our discussion for next changes:

* Named outputs, e.g. `def predict(x: Image) -> Output(y: str, z: str)`
* This is always using parts, which is different from using data.image directly (now, we'll use data.parts[argname].data.image).  That needs to be discussed more broadly as it can make things around monitoring, UI, etc different.
* Need to add name/id field to the Parts proto in the protos repo.
* Provide some example models (either in examples or here) for testing
* We might be better off not supporting unlimited dict nesting levels for now.  However, a json dict might be good to support for params and other definitions.  That may not need to be in this PR though.
* We should support the following python types as "atomics" --- some of these are not yet there, e.g.  bytes or int:
  * str, bytes, int, float, bool, np.ndarray, PIL.Image.Image

zeiler commented on clarifai/runners/models/model_class.py:
```diff
@@ -1,41 +1,232 @@
+    outputs = []
+    inputs = self._convert_proto_to_python(request.inputs)
+    if len(inputs) == 1:
+      inputs = inputs[0]
```
oh this will be nice as in this wrapper we can handle things that the user should not have to deal with like including in each output the input.id too which some of our APIs / code

deigen reviewed (commented):
Looks like this is making progress.  I'm still reading over this revision, but have some comments so far on generate and stream:

* There are a few issues with batched_predict() and batched_generate(), which I mention in my other comments.  Because of these, I think it would be best to remove these functions from this PR and implement them in a subsequent one.
* stream() should be called once with an iterator, not once per input in the stream (see inline comment on this)

deigen commented on clarifai/runners/utils/data_handler.py:
```diff
@@ -1,231 +1,246 @@
+      part.data.image.CopyFrom(image.to_proto())
+    elif isinstance(part_value, list):
+      if len(part_value) == 0:
+        raise ValueError("List must have at least one element")
```
Should be OK with empty list.

zeiler commented on clarifai/runners/utils/data_handler.py:
```diff
@@ -1,231 +1,248 @@
+      else:
+        for part_item in part_value:
+          if isinstance(part_item, Text):
+            part.data.parts.add().data.text.CopyFrom(part_item.to_proto())
```
wouldn't this whole loop just do part = part.data.parts.add()

then use a function that does the same stuff above before the else?

luv-bansal replied:
yeah, refracted this function

zeiler commented on clarifai/runners/utils/data_handler.py:
```diff
@@ -1,231 +1,248 @@
+      kwargs[part_name] = metadata_to_dict(part_data)
+    elif part_data.parts:
+      kwargs[part_name] = []
+      for part_item in part_data.parts:
```
same thing here, re-use a function so each field doesn't have to be updated in two places.

zeiler commented on clarifai/client/model.py:
```diff
@@ -407,47 +408,147 @@ def list_versions(self, page_no: int = None,
-
-      if status_is_retryable(response.status.code) and \
-              time.time() - start_time < 60 * 10:  # 10 minutes
+      if response.status.code == status_code_pb2.MODEL_DEPLOYING and time.time(
```
same thing in a few other places here

luv-bansal replied:
fixed it

luv-bansal commented on clarifai/runners/models/model_class.py:
```diff
@@ -138,56 +140,35 @@ def _convert_part_data(self, data: resources_pb2.Data, param_type: type) -> Any:
+          list_output.append(Audio(part.data.audio))
+        elif part.data.HasField("video"):
+          list_output.append(Video(part.data.video))
+        elif part.data.HasField("bytes_value"):
```
Yeah, you're correct—HasField won't work on built-in types. And when checking for zero values, we can't determine whether a field was not provided or if the user explicitly set it to 0. This is a main concern

luv-bansal replied:
for now I'm checking for zero, but I think we can wrap each of the new fields in a message for proper validation

luv-bansal commented on clarifai/runners/models/model_class.py:
```diff
@@ -1,41 +1,186 @@
+    """Batch predict method for multiple inputs."""
+    with ThreadPoolExecutor() as executor:
+      futures = [executor.submit(self.predict, **input) for input in inputs]
+      return [future.result() for future in futures]
```
For now, I've update this to simple for loop implementation as per your suggestion, we can deal optimising batching in separate PR

luv-bansal commented on clarifai/runners/models/model_class.py:
```diff
@@ -1,41 +1,186 @@
+      inputs = self._convert_proto_to_python(request.inputs)
+      if len(inputs) == 1:
+        inputs = inputs[0]
+        for output in self.stream(**inputs):
```
Oh yeah `Stream_wrapper` function is currently wrong implemented. I'm not sure how to correctly implement it

luv-bansal commented:
@deigen, I've addressed all the points you mentioned in [your comment](https://github.com/Clarifai/clarifai-python/pull/516#issuecomment-2654572280) and everything we discussed on the call.

you specially asked me focus on these two points:

> - Make sure to test with invalid client calls with the wrong types --- What is the error provided to the user in these cases? It should be along the lines of what they would get calling a function with the wrong args.
> - Test with the case where the server defines def predict(x: str) -> str: return x and the client calls with model.predict(x=Image.open('test.jpg'). What happens? Right now I think it will return the empty string (there is nothing in parts[x].text.raw) but it would be better to error on mismatched types. However, if the user actually calls with an empty string, it should still return the empty string for model.predict(x='')

For both cases, the model now throws errors with clear messages:
 
Case 1 (Invalid parameter in predict method):
```
Exception: Model Predict failed with response code: FAILURE
details: "Unknown parameter: `text3` in predict method, available parameters: odict_keys([\'text1\', \'text2\'])"
req_id: "sdk-python-11.1.5-dec9f9995bac4d53ba30cbadac651ae2"
```

Case 2 (Incorrect data type)
```
Exception: Model Predict failed with response code: FAILURE
details: "expected str datatype but the provided input is not a str"
req_id: "sdk-python-11.1.5-f79bd7b9739a4879b9f2abcd774012f8"
```
